### PR TITLE
Add tests for comparator in DiceRoll.java

### PR DESCRIPTION
**Pull request by RoiEXLab**
## Overview
This is one of the methods that causes flaky coverage, the comparator isn't always invoked in a deterministic way, so I wrote tests that would cover all if cases all of the time.

## Functional Changes

[ ] New map or map update
[ ] New Feature
[ ] Feature update or enhancement
[ ] Feature Removal
[x] Code Cleanup or refactor
[ ] Configuration Change
[ ] Bug fix

## Testing

[ ] Covered by existing automated tests
[x] Covered by newly added automated tests
[ ] Manually tested
[ ] No testing done

### CODE REVIEW

File: game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java [hunk 1]
```diff
@@ -741,4 +750,61 @@ void testSbrRolls() {
...
+    @Mock private Unit unit4;
+    @Mock private Unit unit5;
+
+    private final Map<Unit, Tuple<Integer, Integer>> unitPowerAndRollsMap = new HashMap<>();
+    private final List<Unit> units = new ArrayList<>();
+
+    @BeforeEach
+    void setUp() {
+      units.addAll(List.of(unit1, unit2, unit3, unit4, unit5));
+      for (final var unit : units) {
```
Comment (DanVanAtta) [inline game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java]: I think it would help maintainers skip a few steps if this for statement were extracted to a method or a comment is added. Can something be done to make it a bit more direct/obvious what is happening?

File: game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java [hunk 2]
```diff
@@ -741,4 +750,61 @@ void testSbrRolls() {
...
+        when(unit.getType()).thenReturn(unitTypeMock);
+      }
+    }
+
+    @Test
+    void testAttacking() {
+      int index = 4;
+      for (final var unit : units) {
+        when(((UnitAttachment) unit.getType().getAttachment("")).getOffensiveAttackAa(any()))
+            .thenReturn(index / 2);
```
Comment (DanVanAtta) [inline game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java]: What does 'index / 2' represent? Would that question be answered by extracting the value to a variable?

File: game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java [hunk 3]
```diff
@@ -741,4 +750,61 @@ void testSbrRolls() {
...
+      DiceRoll.sortAaHighToLow(units, gameData, false, unitPowerAndRollsMap);
+      assertThat(units.get(0), is(unit1));
+      assertThat(units.get(1), is(unit2));
+      assertThat(units.get(2), is(unit3));
+      assertThat(units.get(3), is(unit4));
+      assertThat(units.get(4), is(unit5));
+    }
+
+    @Test
+    void testDefending() {
```
Comment (DanVanAtta) [inline game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java]: A javadoc comment on how this method is setup to test I think would go a long way. Looking at the loop and comparing to the last test case, it kinda seems like we would get a reverse sort order. Why do we expect the order that is being asserted?

Comment (RoiEXLab) [reply game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java]: I'm pretty sure the sort algorithm used by java is stable, so because multiple values share the same value, they'll stay in their initial order

File: game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java (hunk 2 above)
Comment (RoiEXLab) [reply game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java]: Not really, in order to test the `getFirst() == 0` cases, I had to introduce duplicate keys somehow to get this edge case covered.

A comment would help though IMO

File: game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java (hunk 1 above)
Comment (RoiEXLab) [reply game-core/src/test/java/games/strategy/triplea/delegate/DiceRollTest.java]: Initially I had all the setup in this method, but because only parts of it were used in each test, this was against the strict stubbing policy of mockito, that's why I moved the stuff in such a weird way, I think I can fix that

File: game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java [hunk 4]
```diff
@@ -424,37 +425,19 @@ private static void sortAaHighToLow(
     sortAaHighToLow(units, data, defending, new HashMap<>());
   }
 
-  private static void sortAaHighToLow(
+  @VisibleForTesting
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java]: Instead of making this visible for testing, any thoughts to extract this method to a package-protected utility class, something like "DiceRollSorter"?

File: game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java [hunk 5]
```diff
@@ -424,37 +425,19 @@ private static void sortAaHighToLow(
...
-          }
-          return 0;
-        };
-    units.sort(comparator);
+    units.sort(
+        Comparator.comparing(
+            unit ->
+                getMaxAaAttackAndDiceSides(
+                    Collections.singleton(unit), data, defending, unitPowerAndRollsMap),
+            Comparator.<Tuple<Integer, Integer>, Boolean>comparing(tuple -> tuple.getFirst() == 0)
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java]: A comment to describe the compare algo could be nice to have as well. Maybe extracting to a method would work too so the method name can be used to describe the operation that is occurring.

Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java]: I mean if I understood the algorithm that would be a feasible thing to do, but I don't unfortunately.

File: game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java (hunk 4 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/delegate/DiceRoll.java]: Not feasible for this PR.
The method makes use of `getMaxAaAttackAndDiceSides`, which could be used in a static way, but at this point we'd create a utility class that uses the class it is called from
